Cannot send a concrete type to a chan of an interface

The following program `sample.go` triggers a panic:

```go
package main

import (
    "encoding/json"
    "fmt"
)

type MyJsonMarshaller struct {
}

func (m MyJsonMarshaller) MarshalJSON() ([]byte, error) {
    return []byte("{}"), nil
}

func main() {
    ch := make(chan json.Marshaler, 1)

    marshaller := MyJsonMarshaller{}
    ch <- marshaller
    fmt.Println("OK")
}
```

Expected result:
```console
$ go run ./sample.go
// output
OK
```

Got:
```console
$ yaegi ./sample.go
// output
./main.go:20:5: panic
run: reflect.Value.Send: value of type struct {} is not assignable to type json.Marshaler
goroutine 1 [running]:
runtime/debug.Stack(0x1, 0xc000166a00, 0x40)
	/home/juliens/go-current/src/runtime/debug/stack.go:24 +0x9f
github.com/traefik/yaegi/interp.(*Interpreter).eval.func1(0xc000225c70)
	/home/juliens/dev/go/src/github.com/traefik/yaegi/interp/interp.go:485 +0xc5
panic(0xdc9fa0, 0xc000195050)
	/home/juliens/go-current/src/runtime/panic.go:969 +0x1b9
github.com/traefik/yaegi/interp.runCfg.func1(0xc000132dc0, 0xc0004d4e00, 0xc000225a60)
	/home/juliens/dev/go/src/github.com/traefik/yaegi/interp/run.go:184 +0x247
panic(0xdc9fa0, 0xc000195050)
	/home/juliens/go-current/src/runtime/panic.go:969 +0x1b9
reflect.Value.assignTo(0xdf55e0, 0x1571db8, 0x199, 0xf0fea4, 0x12, 0xe1e2a0, 0x0, 0x0, 0x0, 0x0)
	/home/juliens/go-current/src/reflect/value.go:2425 +0x405
reflect.Value.send(0xc0001b0a80, 0xc00000e840, 0x192, 0xdf55e0, 0x1571db8, 0x199, 0x0, 0x0)
	/home/juliens/go-current/src/reflect/value.go:1534 +0xa6
reflect.Value.Send(0xc0001b0a80, 0xc00000e840, 0x192, 0xdf55e0, 0x1571db8, 0x199)
	/home/juliens/go-current/src/reflect/value.go:1523 +0x93
github.com/traefik/yaegi/interp.send.func2(0xc000132dc0, 0xc0004d2340)
	/home/juliens/dev/go/src/github.com/traefik/yaegi/interp/run.go:3162 +0xb6
github.com/traefik/yaegi/interp.runCfg(0xc0004d4e00, 0xc000132dc0)
	/home/juliens/dev/go/src/github.com/traefik/yaegi/interp/run.go:190 +0x87
github.com/traefik/yaegi/interp.(*Interpreter).run(0xc0003d0000, 0xc0004d4600, 0xc000132b00)
	/home/juliens/dev/go/src/github.com/traefik/yaegi/interp/run.go:121 +0x2b0
github.com/traefik/yaegi/interp.(*Interpreter).eval(0xc0003d0000, 0xc00034c480, 0x165, 0x7fff6c6b62cd, 0x9, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/home/juliens/dev/go/src/github.com/traefik/yaegi/interp/interp.go:577 +0x809
github.com/traefik/yaegi/interp.(*Interpreter).EvalPath(0xc0003d0000, 0x7fff6c6b62cd, 0x9, 0xc00034c300, 0x165, 0x0, 0x0, 0xc000171d40)
	/home/juliens/dev/go/src/github.com/traefik/yaegi/interp/interp.go:399 +0xfa
main.runFile(0xc0003d0000, 0x7fff6c6b62cd, 0x9, 0xc00007cd20, 0x1)
	/home/juliens/dev/go/src/github.com/traefik/yaegi/cmd/yaegi/run.go:123 +0xae
main.run(0xc00000c090, 0x1, 0x1, 0x13, 0x13)
	/home/juliens/dev/go/src/github.com/traefik/yaegi/cmd/yaegi/run.go:89 +0x7e5
main.main()
	/home/juliens/dev/go/src/github.com/traefik/yaegi/cmd/yaegi/yaegi.go:144 +0x415

```


Just for information if I use this code, it works as expected.
```go
package main

import (
    "encoding/json"
    "fmt"
)

type MyJsonMarshaller struct {
}

func (m MyJsonMarshaller) MarshalJSON() ([]byte, error) {
    return []byte("{}"), nil
}

func main() {
    ch := make(chan json.Marshaler, 1)

    marshaller := MyJsonMarshaller{}
    var b json.Marshaler = marshaller
    ch <- b
    fmt.Println("OK")
}
```

--- a/_test/issue-1010.go
+++ b/_test/issue-1010.go
@@ -1,0 +1,22 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+)
+
+type MyJsonMarshaler struct{ n int }
+
+func (m MyJsonMarshaler) MarshalJSON() ([]byte, error) {
+	return []byte(fmt.Sprintf(`{"num": %d}`, m.n)), nil
+}
+
+func main() {
+	ch := make(chan json.Marshaler, 1)
+	ch <- MyJsonMarshaler{2}
+	m, err := json.Marshal(<-ch)
+	fmt.Println(string(m), err)
+}
+
+// Output:
+// {"num":2} <nil>

--- a/interp/run.go
+++ b/interp/run.go
@@ -2392,7 +2392,7 @@ func doComposite(n *node, hasType bool, keyed bool) {
 	l := n.level
 	n.exec = func(f *frame) bltn {
 		typ.mu.Lock()
-		// No need to call zero() as doComposite is only called for a structT
+		// No need to call zero() as doComposite is only called for a structT.
 		a := reflect.New(typ.TypeOf()).Elem()
 		typ.mu.Unlock()
 		for i, v := range values {
@@ -3148,19 +3148,27 @@ func convertConstantValue(n *node) {
 // Write to a channel.
 func send(n *node) {
 	next := getExec(n.tnext)
-	value0 := genValue(n.child[0]) // channel
-	convertLiteralValue(n.child[1], n.child[0].typ.val.TypeOf())
-	value1 := genValue(n.child[1]) // value to send
+	c0, c1 := n.child[0], n.child[1]
+	value0 := genValue(c0) // Send channel.
+	convertLiteralValue(c1, c0.typ.val.TypeOf())
+
+	var value1 func(*frame) reflect.Value // Value to send.
+	switch {
+	case isInterfaceBin(c0.typ.val):
+		value1 = genInterfaceWrapper(c1, c0.typ.val.rtype)
+	default:
+		value1 = genValue(c1)
+	}
 
 	if n.interp.cancelChan {
-		// Cancellable send
+		// Send is cancelable, may have some overhead.
 		n.exec = func(f *frame) bltn {
 			ch, data := value0(f), value1(f)
-			// Fast: send on channel doesn't block
+			// Fast: send on channel doesn't block.
 			if ok := ch.TrySend(data); ok {
 				return next
 			}
-			// Slow: send on channel blocks, allow cancel
+			// Slow: send on channel blocks, allow cancel.
 			f.mutex.RLock()
 			done := f.done
 			f.mutex.RUnlock()
@@ -3172,7 +3180,7 @@ func send(n *node) {
 			return next
 		}
 	} else {
-		// Blocking send (less overhead)
+		// Send is non-cancellable, has the least overhead.
 		n.exec = func(f *frame) bltn {
 			value0(f).Send(value1(f))
 			return next

--- a/interp/run.go
+++ b/interp/run.go
@@ -3160,31 +3160,32 @@ func send(n *node) {
 		value1 = genValue(c1)
 	}
 
-	if n.interp.cancelChan {
-		// Send is cancelable, may have some overhead.
-		n.exec = func(f *frame) bltn {
-			ch, data := value0(f), value1(f)
-			// Fast: send on channel doesn't block.
-			if ok := ch.TrySend(data); ok {
-				return next
-			}
-			// Slow: send on channel blocks, allow cancel.
-			f.mutex.RLock()
-			done := f.done
-			f.mutex.RUnlock()
-
-			chosen, _, _ := reflect.Select([]reflect.SelectCase{done, {Dir: reflect.SelectSend, Chan: ch, Send: data}})
-			if chosen == 0 {
-				return nil
-			}
-			return next
-		}
-	} else {
+	if !n.interp.cancelChan {
 		// Send is non-cancellable, has the least overhead.
 		n.exec = func(f *frame) bltn {
 			value0(f).Send(value1(f))
 			return next
 		}
+		return
+	}
+
+	// Send is cancellable, may have some overhead.
+	n.exec = func(f *frame) bltn {
+		ch, data := value0(f), value1(f)
+		// Fast: send on channel doesn't block.
+		if ok := ch.TrySend(data); ok {
+			return next
+		}
+		// Slow: send on channel blocks, allow cancel.
+		f.mutex.RLock()
+		done := f.done
+		f.mutex.RUnlock()
+
+		chosen, _, _ := reflect.Select([]reflect.SelectCase{done, {Dir: reflect.SelectSend, Chan: ch, Send: data}})
+		if chosen == 0 {
+			return nil
+		}
+		return next
 	}
 }
 

--- a/interp/type.go
+++ b/interp/type.go
@@ -1625,6 +1625,10 @@ func isInterfaceSrc(t *itype) bool {
 	return t.cat == interfaceT || (t.cat == aliasT && isInterfaceSrc(t.val))
 }
 
+func isInterfaceBin(t *itype) bool {
+	return t.cat == valueT && t.rtype.Kind() == reflect.Interface
+}
+
 func isInterface(t *itype) bool {
 	return isInterfaceSrc(t) || t.TypeOf() != nil && t.TypeOf().Kind() == reflect.Interface
 }

